### ISSUE TITLE
LRPolicy.DecayOnValidation causes error if used with SquareLossLayer

### ISSUE THREAD

**Issue by scheidan** [pluskid/Mocha.jl#81]
Great package, thanks for all your work!

I tried to use the `DecayOnValidation` learning rate policy together with a SquareLoss and have two questions/bugs:
- I get the error message `policy not defined`. See this [Gist](https://gist.github.com/scheidan/83026b9f4b5f4c7c7079) for an minimal example. Is my setup correct?
- As I understand the [this code](https://github.com/pluskid/Mocha.jl/blob/3cd3c359c1589afdeb91ded81e48ecd86a258b70/src/solvers.jl#L44), the learning rate would only drop when the loss _decreases_, which is not what we want for a SquareLoss.

**Comment by pluskid** [comment]
@scheidan It seems you forgot to link "this gist". I will show a gist with an example setup.

Yes, currently it only drops when the loss decreases.

**Comment by pluskid** [comment]
See [this gist](https://gist.github.com/pluskid/7c9998b0e9047488d20a) for an example of the setup, though I'm using it for accuracy performance measure.

**Comment by scheidan** [comment]
Thanks for the example!
The [Link](https://gist.github.com/scheidan/83026b9f4b5f4c7c7079) should be working now. The example works fine with if a `AccuracyLayer` is used,but not with `SquareLossLayer`.

**Comment by pluskid** [comment]
@scheidan I just pushed some changes that should fix the issue. Can you also test on your side?

Basically, it seems to be that using of macro evaluations caused some variables to be evaluated at compiling time, but they are not defined yet then.

**Comment by scheidan** [comment]
Thanks for looking into that!

I did a small test (see [here](https://gist.github.com/scheidan/83026b9f4b5f4c7c7079)) and observed the following:

- The error message disappeared.

- Using a `SquareLossLayer`, it always reverts to the first snapshot. I guess because `DecayOnValidation` does not know that in this case a smaller statistic is better (as opposed to accuracy). Maybe we need a option for that in `DecayOnValidation`.

- More on a side note: in `add_coffee_break` using the argument `every_n_epoch=2` has the same effect as setting `every_n_iter=2`.

**Comment by pluskid** [comment]
@scheidan Thanks for providing the test case. I updated `DecayOnValidation` to allow something like this:
```
lr_policy = LRPolicy.DecayOnValidation(0.001, "loss-square-loss", 0.5; higher_better=false)
```
And it should work with square loss cost now.

For the `every_n_epoch=2` issue. That is because batch-size is 500, so 1 epoch is consumed every 2 iterations. However, that still does not match **2** epochs. I will look into this issue. Thanks for spotting this!

**Comment by pluskid** [comment]
@scheidan I think I found the issue. Because you use the same net for training and performance validation. The epoch number get increased during the performance validation process, while the iteration counter does not increase because it is changed only by the solver. It is generally not recommended to use the same net for training and performance validation. The following test case runs as expected.

```julia
use_cuda = false
using Mocha
srand(12345678)
############################################################
# Prepare Random Data
############################################################
srand(1234)
modeldir = "modeltest"
N = 1000
M = 20
P = 10
X = rand(M, N)
W = rand(M, P)
B = rand(P, 1)
Y = (W'*X .+ B)
Y = Y + 0.01*randn(size(Y))
############################################################
# Define network
############################################################
if use_cuda
  backend = GPUBackend()
else
  backend = CPUBackend()
end
init(backend)
data_layer = MemoryDataLayer(batch_size=500, data=Array[X,Y])
weight_layer = InnerProductLayer(name="ip",output_dim=P, tops=[:pred], bottoms=[:data])
loss_layer = SquareLossLayer(name="loss", bottoms=[:pred, :label])
net = Net("TEST", backend, [loss_layer, weight_layer, data_layer])
net2 = Net("TEST", backend, [loss_layer, weight_layer, data_layer])
println(net)
############################################################
# Solve
############################################################
nepochs = 10
lr_policy = LRPolicy.DecayOnValidation(0.001, "loss-square-loss", 0.5; higher_better=false)
params = SolverParameters(regu_coef=0.0005, mom_policy=MomPolicy.Fixed(0.9),
                          max_iter=nepochs, lr_policy=lr_policy,
                          load_from=modeldir)
solver = SGD(params)
setup_coffee_lounge(solver, save_into="$modeldir/statistics.jld", every_n_iter=1000)
val_performance = ValidationPerformance(net2)
add_coffee_break(solver, val_performance, every_n_epoch=2) # same at every_n_iter=2
#add_coffee_break(solver, val_performance, every_n_iter=2000)
## register the listener to get notified on performance validation for lr_policy
setup(params.lr_policy, val_performance, solver)
## add_coffee_break(solver, Snapshot(modeldir), every_n_epoch=2) # same at every_n_iter=2
add_coffee_break(solver, Snapshot(modeldir), every_n_iter=2000)
add_coffee_break(solver, TrainingSummary(show_obj_val=true, show_lr=true), every_n_iter=500)
solve(solver, net)
shutdown(backend)
```

**Comment by scheidan** [comment]
Works great.
